metadata-ingestion/src/datahub/ingestion/source/feast.py

```diff
@@ -25,283 +37,334 @@
-            ),
-            aspects=[],
-        )
+        platform = "unknown"
+        name = "unknown"
+
+        if isinstance(source, FileSource):
+            platform = "file"
+
+            name = source.path.replace("://", ".").replace("/", ".")
+
+        if isinstance(source, BigQuerySource):
+            platform = "bigquery"
+            name = source.table
+
+        if isinstance(source, KafkaSource):
+            platform = "kafka"
+            name = source.kafka_options.topic
+
+        if isinstance(source, KinesisSource):
+            platform = "kinesis"
+            name = (
+                f"{source.kinesis_options.region}:{source.kinesis_options.stream_name}"
+            )
+
+        if isinstance(source, RequestDataSource):
+            platform = "request"
+            name = source.name
 
-        entity_sources = []
+        return platform, name
 
-        if ingest_entity["batch_source"] is not None:
-            entity_sources.append(
+    def _get_data_sources(self, feature_view: FeatureView) -> List[str]:
+        """
+        Get data source URN list.
+        """
+
+        sources = []
+
+        if feature_view.batch_source is not None:
+            batch_source_platform, batch_source_name = self._get_data_source_details(
+                feature_view.batch_source
+            )
+            sources.append(
                 builder.make_dataset_urn(
-                    ingest_entity["batch_source_platform"],
-                    ingest_entity["batch_source_name"],
-                    self.config.env,
+                    batch_source_platform,
+                    batch_source_name,
+                    self.source_config.environment,
                 )
             )
 
-        if ingest_entity["stream_source"] is not None:
-            entity_sources.append(
+        if feature_view.stream_source is not None:
+            stream_source_platform, stream_source_name = self._get_data_source_details(
+                feature_view.stream_source
+            )
+            sources.append(
                 builder.make_dataset_urn(
-                    ingest_entity["stream_source_platform"],
-                    ingest_entity["stream_source_name"],
-                    self.config.env,
+                    stream_source_platform,
+                    stream_source_name,
+                    self.source_config.environment,
                 )
             )
 
-        # append entity name and type
+        return sources
+
+    def _get_entity_workunit(
+        self, feature_view: FeatureView, entity: Entity
+    ) -> MetadataWorkUnit:
+        """
+        Generate an MLPrimaryKey work unit for a Feast entity.
+        """
+
+        feature_view_name = f"{self.feature_store.project}.{feature_view.name}"
+
+        entity_snapshot = MLPrimaryKeySnapshot(
+            urn=builder.make_ml_primary_key_urn(feature_view_name, entity.name),
+            aspects=[StatusClass(removed=False)],
+        )
```

Comment:
I'm not super familiar with Datahub conceptually, but from the Feast side an entity can be the "primary key" into multiple different feature views. Is `MLPrimaryKeySnapshot` specific to a `MLFeatureTable`, in this case a FeatureView?

Reply:
Hey @achals - MlPrimaryKey can be associated with multiple MlFeatureTables. The association happens here:

https://github.com/datahub-project/datahub/blob/master/metadata-models/src/main/pegasus/com/linkedin/ml/metadata/MLFeatureTableProperties.pdl#L51